### ISSUE TITLE
Fedora 34 - keyboard unresponsive after system update

### ISSUE THREAD

**Issue by RedBearAK** [rbreaves/kinto#564]
@rbreaves @prosac 

After a recent round of updates to Fedora 34 on two completely different machines (Acer Aspire and a MacBook), I am experiencing a complete failure of all keyboard input when Kinto is active. At first I thought it was some kind of kernel hardware issue with the most recent kernel, but going back to an older kernel made no difference. No key on the keyboard produces any output, and no media keys or system shortcuts will function until I stop the Kinto service. Then everything is fine. 

The cause of this is unknown. It's strange because the service obviously starts to a certain degree, but something is definitely interfering with the functioning of xkeysnail. 

The GUI log output is not very helpful at all, nor do I see any obvious indication of anything going wrong when I manually run the command from the service file in a terminal. But if I didn't have the menu bar icon working  I'd have to log in via SSH from another machine to kill the service. When it's active I can't even switch to a virtual terminal. The keyboard is just dead. 

Open to any ideas for how to troubleshoot this.

**Comment by RedBearAK** [comment]
@rbreaves 

I decided to try again with Fedora 35 in a VM. Had to install a few different packages like `python3-wheel` and `python3-evdev` in order to even get through the Kinto installer, which was strange. Got the "unit not found" error that was happening before the installer was "fixed" for Fedora last year. 

Instead of trying to fix that in the usual way by changing SELinux, I tried Lenbok's `uinput` group trick to let `xkeysnail` run as user. After getting past a strange bug where I had "GNOME with Xorg" selected at the login screen but it would somehow default back to Wayland after a reboot, Kinto running as user seems to work perfectly. 

So changing the Kinto installer to create the `uinput` group, ask the user to restart, and then run `xkeysnail` as user appears to be a viable solution to getting Kinto running again on Fedora. I never found a solution or even an explanation to why Kinto completely stopped working on Fedora last year. But that issue seems to only apply to running the service as root.

**Comment by RedBearAK** [comment]
@rbreaves 

Having the same issue on Fedora with Alt+Tab that I was having with CutefishOS, where it just bounces back and forth between the last two windows that had focus. Like what Alt+Grave was doing on Budgie. 

Just taking a wild guess here, but could it be that this has something to do with running these in a GNOME Boxes VM? Only way to test is to install on bare metal and see what happens, but I'm trying to think if it's even possible that has something to do with the issue. 

Mapping is off while using Boxes, so that it works within the VM, but maybe Alt+Tab is a special case and is having a weird interaction with Boxes?

**Comment by RedBearAK** [comment]
@rbreaves 

Installed on bare metal on my old 2007 MacBook, the problem remains. 

Both Alt+Tab and Alt+Grave are acting like your "held keys" patch to `xkeysnail` is no longer working. 

I'm installing from the Kinto.sh installer URL, and I don't think I was attempting the run-as-user trick with CutefishOS, so it seems this may be a fundamental issue with the current state of the `xkeysnail` that is getting installed with Kinto on a fresh system. 

If this affects all distro variants and I had to do a fresh install right now on a system that I actually intended to use, this issue would make Kinto almost unusable for me.

**Comment by RedBearAK** [comment]
@rbreaves 

I should note that the task switcher dialog is also not appearing on Fedora, either in the VM or bare metal. This same symptom was happening in CutefishOS. If this really is a "held keys" issue, that is probably just down to the keystroke exiting before the short delay that would bring up the task switcher dialog. I know at least GNOME is currently designed with a short delay before it shows the switcher dialog, because there are a couple of GNOME extensions available to disable that delay. 

Tried to install from an earlier release of Kinto, but I don't think that did any good because the installer is cloning the current version of `xkeysnail`. I'll have to try and manually install an earlier version of `xkeysnail` and see if that fixes this issue. 

Here on my main machine, which was just set up from scratch a few weeks ago, I still have `xkeysnail` v0.3.0, but the version that is being installed now is v0.4.0. I am pretty sure at this point that the "held keys" functionality is currently broken. 

I can't think of any other shortcut at the moment that would behave differently with "held keys", so I'm not sure how to provide more evidence that "held keys" is broken.

**Comment by RedBearAK** [comment]
@rbreaves 

I went and tested Fedora 34 again (the version I was using last year on a 2007 MacBook and an Acer Laptop, where I had the original problem in this thread), in a new VM. Kinto installed right after the initial Fedora 34 install, before any system updates, worked fine. 

After doing the first 1.2GB system update and rebooting, the entire keyboard was once again unresponsive (just as I experienced last year), and without access to a working tray icon or the Kinto GUI through using the mouse to get to the Activities Overview, there would be no apparent way to get keyboard control back. 

However, after stopping the service through the GUI (via the mouse and the Activities Overview button and the Applications button on the GNOME Dash) and then **_re-running_** the Kinto setup, Kinto suddenly started working again. And this is persisting even after rebooting the VM multiple times. I can't get the unresponsive keyboard issue to come back. 

So perhaps the secret is that something gets overwritten or changed to a different version when doing some Fedora update that came out sometime mid-2021, and re-running the Kinto setup resets it. 

I'm sure switching to running Kinto as user would also bypass this issue, but that doesn't seem to be necessary if the user can just re-run the Kinto setup after this mystery Fedora update. 

I still don't know what the actual problem was that led to the unresponsive keyboard, but this kind of narrows it down to something that Kinto setup would be capable of fixing by just doing what it normally does. 

So that's good news, I guess. Except that if that same update somehow runs again (because the system notices a version mismatch, or there is an even newer version of whatever that update was) then the user will once again have complete loss of keyboard control, and need to re-run Kinto setup to fix it. 

If Kinto is switched to running as user, it would theoretically be far less likely that this would ever be an issue again. Because that seems to sidestep whatever was causing the problem. But I can't be absolutely sure of that either.
